### PR TITLE
Handle case where attempted LVM2 Thin device reads fail

### PR DESCRIPTION
**Pull request by movitto**
As it stands if a client invokes a disk read against a LVM2 Thin Provisioned
volume for which space has not yet been allocated for the address being read,
a rather confusing / ugly error will occur:

  LVM2Thin cannot find device block: N (closest: X)]

  Where N is the data block corresponding to the address being looked up
  and X is the last allocated data block

This patch catches this case and simply returns an empty string if the read
cannot be completed

### CODE REVIEW

**Comment by movitto** [comment]
@roliveri please review

File: lib/VolumeManager/LVM/lvm2disk.rb [hunk 1]
```diff
@@ -42,7 +42,12 @@ def d_read(pos, len)
     if logicalVolume.thin?
       device_id = logicalVolume.thin_segment.device_id
       thin_pool = logicalVolume.thin_pool_volume
-      data_blks = thin_pool.metadata_volume.superblock.device_to_data(device_id, pos, len)
+      begin
+        data_blks = thin_pool.metadata_volume.superblock.device_to_data(device_id, pos, len)
+      rescue
+        return retStr
```
Comment (roliveri) [inline lib/VolumeManager/LVM/lvm2disk.rb]: @movitto Should it return an empty string, or a string of 0x0 of the requested length?
That is, unless the length causes the read to exceed the end of the device, then it should return a string of 0x0
of the available length.

Reads from a block device should only return an empty string when at the end of the device. I would think LVM2 would maintain that behavior.

**Comment by movitto** [comment]
@roliveri addressed your feedback. Now the lvm2thin module will fill in the remaining length with zero's. So for example if the invoker tries to read 512 bytes starting at 256 bytes from the end of the allocated space they will get the last 256 bytes of data plus 256 0's

As it stands, the lvm2thin d_read method will return zeros if the user requests a read beyond the total lvm volume size but I believe this edge case is handled elsewhere correct? (in the general case where the user tries to read beyond the extent of a disk boundary)

**Comment by roliveri** [comment]
@movitto Right, I believe the `MiqDisk` will check the boundary  of the underlying device.

**Comment by roliveri** [comment]
@movitto What if the unallocated space falls within the extents of a read?
It is possible to read say, 3 blocks, and the second block is unallocated.

So, we need to check for unallocated data on a per-block basis, filling with zeros
as needed. Unallocated data can be encountered at the beginning, middle or end of a read.

**Comment by movitto** [comment]
@roliveri updated, the lvm2thin module now return zeros for all unallocated blocks

File: lib/VolumeManager/LVM/lvm2disk.rb [hunk 2]
```diff
@@ -43,9 +43,17 @@ def d_read(pos, len)
...
       thin_pool = logicalVolume.thin_pool_volume
       data_blks = thin_pool.metadata_volume.superblock.device_to_data(device_id, pos, len)
-      data_blks.each do |offset, len|
-        thin_pool.data_volume.disk.seek offset
-        retStr << thin_pool.data_volume.disk.read(len)
+
+      data_blks.each do |device_blk, data_blk, blk_offset, blk_len|
+        if data_blk.nil?
+          # fill in unallactored data
+          puts blk_len
```
Comment (roliveri) [inline lib/VolumeManager/LVM/lvm2disk.rb]: Remove the `puts`.

File: lib/VolumeManager/LVM/lvm2disk.rb [hunk 3]
```diff
@@ -43,9 +43,17 @@ def d_read(pos, len)
       device_id = logicalVolume.thin_segment.device_id
       thin_pool = logicalVolume.thin_pool_volume
       data_blks = thin_pool.metadata_volume.superblock.device_to_data(device_id, pos, len)
-      data_blks.each do |offset, len|
-        thin_pool.data_volume.disk.seek offset
-        retStr << thin_pool.data_volume.disk.read(len)
+
+      data_blks.each do |device_blk, data_blk, blk_offset, blk_len|
```
Comment (roliveri) [inline lib/VolumeManager/LVM/lvm2disk.rb]: Change `device_blk` to `_device_blk` to quiet code climate.

**Comment by movitto** [comment]
@roliveri updated,

**Comment by simaishi** [comment]
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1488378

@movitto  Is it ok to backport this to manageiq-gems-pending Fine branch?

**Comment by movitto** [comment]
@simaishi hey Satoe, this one go lost somewhere in the gaps. I don't see any reason why this fix could not be backported to previous releases. Go for it.

**Comment by simaishi** [comment]
Fine backport (to manageiq-gems-pending repo) details:

```
$ git log -1
commit fa86b88c418c00fece26db84ccffc823068cf81c
Author: Richard Oliveri <oliveri.richard.github@gmail.com>
Date:   Fri Sep 22 11:44:43 2017 -0400

    Merge pull request #28 from movitto/preallocated_lvm2thin_fix
    
    Handle case where attempted LVM2 Thin device reads fail
    (cherry picked from commit 9ae19959eee351a71919996464bca31bea45c336)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1511147
```

**Comment by simaishi** [comment]
@miq-bot add_label fine/backported

**Pull request merged**